control.lua

```diff
@@ -205,103 +328,46 @@ end
-    end
-    for _, station in pairs(stations) do
-      table.insert(global.stations[name], station)
-    end
-  end
-end
-script.on_nth_tick(300, update_list_stations)
-
-
--- Build list of stations
-function build_list_stations()
-  global.stations = {}
-  local stations = game.surfaces["nauvis"].find_entities_filtered{type= "train-stop"}
-  for _, station in pairs(stations) do
-    if not global.stations[station.backer_name] then
-      global.stations[station.backer_name] = {}
-    end
-    table.insert(global.stations[station.backer_name], station)
-  end
-  debug("Stations list rebuilt")
-end
-
-
 -- Reset train schedule after a train has reached its destination
 function reset_station(id)
-  local records = global.dispatched[id].train.schedule.records
-  local current = global.dispatched[id].train.schedule.current
-
-  -- Only reset the train schedule if it has reached the correct station
-  if records[global.dispatched[id].current] ~= nil and records[global.dispatched[id].current].station == global.dispatched[id].station then
-
-    -- Remove destination station from schedule
-    table.remove(records, global.dispatched[id].current)
+  -- If new train has no schedule, do not give it one and remove it from dispatching list.
+  if global.dispatched[id].train.schedule then
+    local records = global.dispatched[id].train.schedule.records
+    local current = global.dispatched[id].train.schedule.current
+    
+    -- Only reset the train schedule if it has reached the correct station
+    if records[global.dispatched[id].current] ~= nil and records[global.dispatched[id].current].station == global.dispatched[id].station then
+
+      -- Remove destination station from schedule
+      table.remove(records, global.dispatched[id].current)
+
+      -- If the current station is after the destination station
+      if current > global.dispatched[id].current then
+        current = current - 1
+      end
 
-    -- If the current station is after the destination station
-    if current > global.dispatched[id].current then
-      current = current - 1
+      -- Reset train schedule
+      global.dispatched[id].train.schedule = {current=current, records=records}
+      if count_locos(global.dispatched[id].train) > 0 then
+        global.dispatched[id].train.manual_mode = false
+      else
+        global.dispatched[id].train.manual_mode = true
+      end
     end
-
-    -- Reset train schedule
-    global.dispatched[id].train.schedule = {current=current, records=records}
-    global.dispatched[id].train.manual_mode = false
   end
-
+  
   global.dispatched[id] = nil
 end
 
-
--- Get red and green signal for given entity and signal
-function get_signal(entity, signal)
-  local red = entity.get_circuit_network(defines.wire_type.red)
-  local green = entity.get_circuit_network(defines.wire_type.green)
-  local value = nil
-  if red then
-    value = red.get_signal(signal)
+-- Count the number of locomotives in the train
+function count_locos(train)
```

Comment:
You do not want to know the number of locomotives, but only if there are any. Best to rename the function `has_loco`, and change the return value to `true` or `false`.

Reply:
You are right.  I copied this from another mod, fixed it in that mod, and forgot to fix it here.